### ISSUE TITLE
procedure GetToken in PSQLTypes.pas does not parse well '::'

### ISSUE THREAD

**Issue by macc2010** [pashagolub/postgresdac#20]
Hello, 

I have seen that when I execute a sql expression in TPSqlDatabase that contains : 

ROW( 1, 2, 3 ) :: type_mytype

In the TNativeDataSet.QuerySetParams the procedure GetToken is called and it converts the double colon '::' in simple colon ':' resulting in this expression : 

ROW( 1, 2, 3 ) : type_mytype

**Comment by macc2010** [comment]
I have changed the code of GetToken procedure and it now pase well the double colon : 

´´´
procedure GetToken(var Buffer, Token: string);
label ExitProc;
var
  P, nLen: Integer;
  Quote: string;
begin
  P := {$IFNDEF NEXTGEN}1{$ELSE}0{$ENDIF};
  nLen := Length(Buffer);
  Token  := '';
  if Buffer = '' then Exit;
  {$IFNDEF NEXTGEN}
  while CharInSet(Buffer[P], [' ',#9]) do
  {$ELSE}
  while Buffer[P].IsInArray([' ',#9]) do
  {$ENDIF}
  begin
    Inc(P);
    //if Length(Buffer) < P then  goto ExitProc;
    if nLen < P then goto ExitProc;
  end;
  if (Pos(Buffer[P],DELIMITERS) <> 0) then
  begin
    // if Buffer[P]+Buffer[P+1] is '::', for example when processing an expression like this ROW( 1, 2, 3 ) :: type_x
    if ( Buffer[P] = ':' ) and ( P+1 < nLen ) and ( Buffer[P+1] = ':' ) then
    begin
      Token := Buffer[P] + Buffer[P+1];
      P := P + 2;
      goto ExitProc;
    end
    else
    begin
      Token  := Buffer[P];
      Inc(P);
      goto ExitProc;
    end;
  end;
  {$IFNDEF NEXTGEN}
  if CharInSet(Buffer[P], ['"','''']) then
  {$ELSE}
  if Buffer[P].IsInArray(['"','''']) then
  {$ENDIF}
  begin
    Quote  := Buffer[P];
    Token  := Quote;
    Inc(P);
    //while P <= Length(Buffer) do
    while P <= nLen do
    begin
      Token := Token + Buffer[P];
      Inc(P);
      if (Buffer[P-1] = Quote) and (Buffer[P-2] <> '\') then  Break;
    end;
  end else
  begin
    //while P <= Length(Buffer) do
    while P <= nLen do
    begin
      Token := Token + Buffer[P];
      Inc(P);
      //if (P > Length(Buffer)) or (Pos(Buffer[P],DELIMITERS) <> 0)
      if (P > nLen) or (Pos(Buffer[P],DELIMITERS) <> 0)
        {$IFNDEF NEXTGEN}
        or CharInSet(Buffer[P], ['"','''']) then Break;
        {$ELSE}
        or Buffer[P].IsInArray(['"','''']) then Break;
        {$ENDIF}
    end;
  end;
ExitProc:
  Delete(Buffer, 1, {$IFNDEF NEXTGEN}P-1{$ELSE}P{$ENDIF});
end;
´´´

**Comment by pashagolub** [comment]
That's a expected behaviour. Set `TPSQLQuery.ParamCheck` to False if you don't want automatic parameter creation.

**Comment by macc2010** [comment]
Hello,

I am using a stored procedure with more than 100 parameters, but as Postgresql does not allow more than 100 parameters in a stored procedure I have used an argument with of composite type. To avoid problems, I need to cast the parameters with a ROW and with double colon before the type of the parameter, the problem is that I need to ejecute the TNativeDataSet.QuerySetParams and if I set ParamCheck to False, the parse is not executed

Aparently, the TNativeDataSet.QuerySetParams shouldn't never parse the double colon '::' but it does, the objetive of the simple colon in SQL properties in Delphi is to be able to replace the identifier passed next to the colon by a valid sql value. But there is an exception, if after a colon, there is another colon, the second colon itself is not a valid identifier for a parameter, so I do not understand why the double colon is not ignored by the GetToken procedure

Is there any case that could affect this modification to the parse of parameters before the execution of existings DML?

Finally, if you see my version of the GetToken procedure, there is also an optimization to not to have to calculate the Length(Buffer) several times in the code, it is calculated once though the use of the nLen variable and used this variable in the code all the times that the Length(Buffer) is needed, increasing the performance of the code

Sorry if I make questions that are trivial, sorry by the inconvenience

I like a lot your code

Thank you very much
